d2ir/compile.go

```diff
@@ -92,7 +93,114 @@ func (c *compiler) compileClasses(m *Map) {
 			l.Fields = append(l.Fields, base)
 		}
 
-		c.compileClasses(l)
+		c.overlayClasses(l)
+	}
+}
+
+func (c *compiler) resolveSubstitutions(refctx *RefContext) {
+	boardScope := refctx.ScopeMap
+	if NodeBoardKind(refctx.ScopeMap) == "" {
+		boardScope = ParentBoard(refctx.ScopeMap).Map()
+	}
+
+	varsMap := &Map{}
+	vars := boardScope.GetField("vars")
+	if vars != nil {
+		varsMap = vars.Map()
+	}
+
+	subbed := false
+	switch {
+	case refctx.Key.Value.UnquotedString != nil:
+		for i, box := range refctx.Key.Value.UnquotedString.Value {
+			if box.Substitution != nil {
+				resolvedField := c.resolveSubstitution(varsMap, refctx.Key, box.Substitution)
+				if resolvedField != nil {
+					// If lone and unquoted, replace with value of sub
+					if len(refctx.Key.Value.UnquotedString.Value) == 1 {
+						refctx.Key.Value = d2ast.MakeValueBox(resolvedField.Primary().Value)
+					} else {
+						refctx.Key.Value.UnquotedString.Value[i].String = go2.Pointer(resolvedField.Primary().String())
+						subbed = true
+					}
+				}
+			}
+		}
+		if subbed {
+			refctx.Key.Value.UnquotedString.Coalesce()
+		}
+	case refctx.Key.Value.DoubleQuotedString != nil:
+		for i, box := range refctx.Key.Value.DoubleQuotedString.Value {
+			if box.Substitution != nil {
+				resolvedField := c.resolveSubstitution(varsMap, refctx.Key, box.Substitution)
+				if resolvedField != nil {
+					refctx.Key.Value.DoubleQuotedString.Value[i].String = go2.Pointer(resolvedField.Primary().String())
+					subbed = true
+				}
+			}
+		}
+		if subbed {
+			refctx.Key.Value.DoubleQuotedString.Coalesce()
+		}
+	}
+}
+
+func (c *compiler) resolveSubstitution(vars *Map, mk *d2ast.Key, substitution *d2ast.Substitution) *Field {
+	var resolved *Field
+	for _, p := range substitution.Path {
+		if vars == nil {
+			resolved = nil
+			break
+		}
+		r := vars.GetField(p.Unbox().ScalarString())
+		if r == nil {
+			resolved = nil
+			break
+		}
+		vars = r.Map()
+		resolved = r
+	}
+
+	if resolved == nil {
+		c.errorf(mk, `could not resolve variable "%s"`, strings.Join(substitution.IDA(), "."))
+	} else if resolved.Composite != nil {
+		c.errorf(mk, `cannot reference map variable "%s"`, strings.Join(substitution.IDA(), "."))
+	} else {
+		return resolved
+	}
+	return nil
+}
+
+func (c *compiler) compileVars(m *Map, ast *d2ast.Map) {
+	for _, n := range ast.Nodes {
+		if n.MapKey != nil && n.MapKey.Key != nil && len(n.MapKey.Key.Path) == 1 && strings.EqualFold(n.MapKey.Key.Path[0].Unbox().ScalarString(), "vars") {
+			c.compileKey(&RefContext{
+				Key:      n.MapKey,
+				Scope:    ast,
+				ScopeMap: m,
+				ScopeAST: ast,
+			})
+			break
```

Comment:
You could have more than one vars map in a map, so you can't break here.

Reply:
Either way it's best you do resolution of variables all at the end so you don't need any of this code.